src/aal/aal_arm.h

```diff
@@ -0,0 +1,53 @@
+#pragma once
+
+#if defined(__arch64__)
+#  define SNMALLOC_VA_BITS_64
+#else
+#  define SNMALLOC_VA_BITS_32
+#endif
+
+#include <iostream>
+namespace snmalloc
+{
+  /**
+   * ARM-specific architecture abstraction layer.
+   */
+  class AAL_arm
+  {
+  public:
+    /**
+     * Bitmap of AalFeature flags
+     */
+    static constexpr uint64_t aal_features = IntegerPointers;
+    /**
+     * On pipelined processors, notify the core that we are in a spin loop and
+     * that speculative execution past this point may not be a performance gain.
+     */
+    static inline void pause()
+    {
+      __asm__ volatile("yield");
+    }
+
+    /**
+     * Issue a prefetch hint at the specified address.
+     */
+    static inline void prefetch(void* ptr)
+    {
+      __builtin_prefetch(ptr, 0, 1);
+    }
+
+    /**
+     * Return a cycle counter value.
+     * on ARM cpu counters are accessible only in privileged mode
+     */
+    static inline uint64_t tick()
+    {
+      struct timespec n = {0, 0ul};
+      clock_gettime(CLOCK_MONOTONIC, &n);
+
+      return static_cast<uint64_t>((n.tv_sec) * (1000000000ul * n.tv_nsec));
```

Comment:
Given that this is a fallback path, it should probably be in the generic version.

It looks as if the version in `AAL_Generic` is will unconditionally use `__builtin_readcyclecounter` if it exists, so we probably need to make that conditional on the builtin supporting a in unprivileged mode, which probably merits a new AAL feature flag.

Reply:
This is not exactly the right place to ask, but it seems roughly on topic: is there guidance for when to use a AAL/PAL feature flag vs. inheritance and overriding?

Reply:
For the AAL, it's relatively simple: the per-arch version is more-or-less a mixin and so you can't override the generic behaviour from it, you can only provide feature flags that tell the generic version when to call into the arch-specific one.  We should probably be using a method implementation check trait to determine whether to call the arch version or use a builtin, rather than always using the builtin where available.

Note that the AAL is currently an abstraction over one thing and so is almost certainly the wrong interface.  I'm very happy to accept PRs that improve it!  It was intended as a starting point so that we could figure out what the real design should look like when we had 2-3 examples (so please do submit a PR for the MIPS one!).

For the PAL, the rule of thumb is that we should override things when we want to provide a complete implementation of something that's different on all platforms, but use feature flags when we want to have most of a method shared by provide extension hooks, or mechanism for skipping something, or to allow callers to conditionally call a particular function only if it exists.  For example, we use the `AlignedAllocation` feature flag to indicate to callers that they can call the function that requests very strongly aligned memory, `LowMemoryNotification` to allow callers to register to receive low memory notifications to lazily decommit memory if memory is low, and `LazyCommit` to determine whether we should allocate a large contiguous virtual address range for the pagemap and expect it to be backed by writealbe zeroed pages on demand. 

The `AAL_Generic` class is also used to define the interface for the AAL.  Methods on the `AAL_Arch` class are intended to be used only via the `AAL_Generic` class.  I'm not 100% sure this is the right design, but it seemed like a good idea at the time...